crates/solver-service/src/apis/order_validation.rs

```diff
@@ -0,0 +1,550 @@
+use alloy_primitives::{hex, Address as AlloyAddress, U256};
+use alloy_sol_types::SolCall;
+use solver_config::Config;
+use solver_core::SolverEngine;
+use solver_types::{
+	standards::eip7683::{interfaces::StandardOrder, LockType},
+	APIError, ApiErrorType,
+};
+use std::convert::TryFrom;
+
+mod interfaces {
+	use alloy_sol_types::sol;
+
+	sol! {
+		function balanceOf(address owner, uint256 id) external view returns (uint256);
+	}
+}
+
+pub async fn ensure_user_capacity_for_order(
+	solver: &SolverEngine,
+	config: &Config,
+	lock_type: LockType,
+	standard_order: &StandardOrder,
+) -> Result<(), APIError> {
+	let origin_chain_id: u64 =
+		u64::try_from(standard_order.originChainId).map_err(|_| APIError::BadRequest {
+			error_type: ApiErrorType::OrderValidationFailed,
+			message: "Origin chain ID missing or invalid in order".to_string(),
+			details: None,
+		})?;
+
+	let user = &standard_order.user;
+
+	match lock_type {
+		LockType::ResourceLock => {
+			for input in &standard_order.inputs {
+				let token_id = input[0];
+				let amount = input[1];
+				if amount.is_zero() {
+					continue;
+				}
+				validate_compact_deposit_for_order(
+					solver,
+					config,
+					origin_chain_id,
+					user,
+					token_id,
+					amount,
+				)
+				.await?;
+			}
+		},
+		_ => {
+			//  double check because when executing the openOrder will fail.
+			// We can reduce RPC Calls if we delete this validation
+			// But UX increase since will detect earlier if user has insufficient balance.
```

Comment:
```suggestion
			// Defensive check: validates sufficient balance before calling `openFor`.
			// Trade-off: This adds an extra RPC call, but improves UX by failing fast
			// with a clear error rather than waiting for the `openFor` transaction to revert.
			// The `openFor` call would fail anyway with insufficient balance, so this
			// check is optional but recommended for better user experience.
```